src/execution/one-shot.ts

```diff
@@ -180,158 +202,97 @@ export class OneShotExecution extends BaseExecution<OneShotScriptConfig> {
+    if (!fingerprint.data.cacheable) {
+      return false;
     }
+    const prevFingerprint = await this.#readPreviousFingerprint();
+    return prevFingerprint !== undefined && fingerprint.equal(prevFingerprint);
+  }
 
-    // Computing the fingerprint can take some time, and the next operation is
-    // destructive. Another good opportunity to check if we should still start.
-    if (this.#shouldNotStart) {
-      return {ok: false, error: [this.#startCancelledEvent]};
-    }
+  /**
+   * Handle the outcome where the script is already fresh.
+   */
+  async #handleFresh(fingerprint: Fingerprint): Promise<ExecutionResult> {
+    // TODO(aomarks) Does not preserve original order of stdout vs stderr
+    // chunks. See https://github.com/google/wireit/issues/74.
+    await Promise.all([
+      this.#replayStdoutIfPresent(),
+      this.#replayStderrIfPresent(),
+    ]);
+    this.logger.log({
+      script: this.script,
+      type: 'success',
+      reason: 'fresh',
+    });
+    return {ok: true, value: fingerprint};
+  }
+
+  /**
+   * Handle the outcome where the script was stale and we got a cache hit.
+   */
+  async #handleCacheHit(
+    cacheHit: CacheHit,
+    fingerprint: Fingerprint
+  ): Promise<ExecutionResult> {
+    await this.#prepareDataDir();
+
+    // If we are restoring from cache, we should always delete existing output.
+    // The purpose of "clean:false" and "clean:if-file-deleted" is to allow
+    // tools with incremental build (like tsc --build) to work.
+    //
+    // However, this only applies when the tool is able to observe each
+    // incremental change to the input files. When we restore from cache, we are
+    // directly replacing the output files, and not invoking the tool at all, so
+    // there is no way for the tool to do any cleanup.
+    await this.#cleanOutput();
+
+    await cacheHit.apply();
+
+    // We include stdout and stderr replay files when we save to the cache, so
+    // if there were any, they will now be in place.
+    // TODO(aomarks) Does not preserve original order of stdout vs stderr
+    // chunks. See https://github.com/google/wireit/issues/74.
+    await Promise.all([
+      this.#replayStdoutIfPresent(),
+      this.#replayStderrIfPresent(),
+    ]);
+
+    await this.#writeFingerprintFile(fingerprint);
+
+    this.logger.log({
+      script: this.script,
+      type: 'success',
+      reason: 'cached',
+    });
+
+    return {ok: true, value: fingerprint};
+  }
+
+  /**
+   * Handle the outcome where the script was stale and we need to run it.
+   */
+  async #handleNeedsRun(fingerprint: Fingerprint): Promise<ExecutionResult> {
+    const shouldClean = await this.#shouldClean(fingerprint);
 
     // It's important that we delete the previous fingerprint and stdio replay
     // files before running the command or restoring from cache, because if
     // either fails mid-flight, we don't want to think that the previous
     // fingerprint is still valid.
```

Comment:
The "or restoring from cache" portion of this comment doesn't belong here now, but should the same note be added above the `await this.#prepareDataDir();` inside `#handleCacheHit()`?